### DISCUSSION TITLE
How to do tree-shaking of aws-sdk?

### DISCUSSION THREAD

**Discussion by Kelvijn** [mrgrain/cdk-esbuild#113] [Q&A] [answered]
Dear devs,

How to use tree shaking?
I tried with TransformOptions but I cannot import this from the '@mrgrain/cdk-esbuild' package, which I found in https://constructs.dev/packages/@mrgrain/cdk-esbuild/v/2.0.0-alpha.1#mrgraincdkesbuildtransformoptions

My lambda currently is 500k lines long.

Kind regards,
Kelvijn

**Comment by mrgrain** [comment]
`TransformOptions` is only used for inline code, which only allows 4kb of code. You probably want to use `TypeScriptCode` or `JavaScriptCode` which uses `BuildOptions`.

According to [esbuild docs](https://esbuild.github.io/api/#tree-shaking), tree-shaking is only enabled either when bundling is enabled or when the output format is set to iife. But can be force-enabled by setting the respective option:

```ts
import * as lambda from "@aws-cdk/aws-lambda";
import { TypeScriptCode } from "@mrgrain/cdk-esbuild";

const bundledCode = new TypeScriptCode("src/index.ts", {
  buildOptions: {
    bundle: true, // or
    treeShaking: true,
  }
});

const fn = new lambda.Function(stack, "MyFunction", {
  runtime: lambda.Runtime.NODEJS_14_X,
  handler: "index.handler",
  code: bundledCode,
});
```

Please let me know if that works for you!


It also should work with `InlineTypeScriptCode`, if you really have to:
```ts
new InlineTypeScriptCode("const code: number = 1 + 2;", {
  buildOptions: {
    treeShaking: true,
  },
});
```

See:
https://constructs.dev/packages/@mrgrain/cdk-esbuild/v/2.0.0-alpha.1#mrgraincdkesbuildbuildoptionspropertybundle
https://constructs.dev/packages/@mrgrain/cdk-esbuild/v/2.0.0-alpha.1#mrgraincdkesbuildbuildoptionspropertytreeshaking

**Reply by Kelvijn** [reply]
Thank you for your quick response!!!

I will check later today or tomorrow.

**Reply by Kelvijn** [reply]
My cdk.out assets for the lambdas are still huge. (I also tried both options seperatly) 
I have one package.json that all of my lambda's use.

My imports for CreateSession.ts :
import { v4 as uuidv4 } from 'uuid'
import * as AWS from 'aws-sdk'
import { makeHandler } from '../utils'


<img width="1080" alt="Screenshot 2021-10-17 at 20 02 58" src="https://user-images.githubusercontent.com/90686663/137639550-c56d3386-2c3d-4378-b439-aa9e2b863fb5.png">

<img width="1080" alt="Screenshot 2021-10-17 at 20 02 41" src="https://user-images.githubusercontent.com/90686663/137639552-3f27711a-7a57-47c2-a6f1-f7ee663ca2bc.png">

I haven't looked into lambda layers yet, because now all my node modules are packages in every lambda and maybe it is not possible to apply treeshaking in this setup.

**Reply by mrgrain** [reply]
Ah yes, I see. This will be because  of this: `import * as AWS from 'aws-sdk'`

Long story short, esbuild doesn't know what parts of the SDK to "shake". You will have to change how you import `aws-sdk`. AWS has written a good blog post about this: https://aws.amazon.com/blogs/developer/modular-packages-in-aws-sdk-for-javascript/

E.g. to use S3 you will have to write
```ts
import S3 from "aws-sdk/clients/s3";
```
instead.

---

Alternatively, because you're creating a Lambda, you can mark `aws-sdk` as `external` und fallback to the version that comes pre-installed on Lambda:

```ts
new TypeScriptCode("./src/index.tsx", {
  buildOptions: {
    external: ["aws-sdk"],
  },
});
```

**Reply by Kelvijn** [reply]
I reduced the size of my lambdas to 1/10th by using the new V3 modular API that you suggested.

Thanks!!

Also great job with this library, makes bundling lambdas EZ!

**Comment by mrgrain** [comment, marked as answer]
Ah yes, I see. This will be because  of this: `import * as AWS from 'aws-sdk'`

Long story short, esbuild doesn't know what parts of the SDK to "shake". You will have to change how you import `aws-sdk`. AWS has written a good blog post about this: https://aws.amazon.com/blogs/developer/modular-packages-in-aws-sdk-for-javascript/

E.g. to use S3 you will have to write
```ts
import S3 from "aws-sdk/clients/s3";
```
instead.

---

Alternatively, because you're creating a Lambda, you can mark `aws-sdk` as `external` und fallback to the version that comes pre-installed on Lambda:

```ts
new TypeScriptCode("./src/index.tsx", {
  buildOptions: {
    external: ["aws-sdk"],
  },
});
```
